### PR TITLE
Add hash and signature validation

### PR DESCRIPTION
**Pull request by tobiaskohlbau**
Hi,

thanks for your work on nice update experience with go programs.
Selfupdate should support the functionality to check/validate downloaded files in advance.
This PR adds support for validation by introducing a `Validator` interface. The user can implement own
validation logic or use the supplied samples of SHA2 and ECDSA.

I would appreciate if you have time to look into this and let me know if there are any changes required.

Best regards
Tobias

### CODE REVIEW

**Comment by rhysd** [comment]
@tobiaskohlbau Could you please describe how to put signature files for each versions and platforms in README.md also?

File: selfupdate/detect.go [hunk 1]
```diff
@@ -59,6 +59,16 @@ func findAssetFromReleasse(rel *github.RepositoryRelease, suffixes []string, tar
 	return nil, semver.Version{}, false
 }
 
+func findValidationAsset(rel *github.RepositoryRelease, validationName string) (*github.ReleaseAsset, bool) {
+	for _, asset := range rel.Assets {
+		name := asset.GetName()
+		if name == validationName {
```
[the remark this answers is not in the document]
Comment (tobiaskohlbau) [reply selfupdate/detect.go]: Good catch, what about dropping variable name?
`if validationName == asset.GetName() {`

File: selfupdate/update.go [hunk 2]
```diff
@@ -13,8 +15,10 @@ import (
 	"github.com/inconshreveable/go-update"
 )
 
-func uncompressAndUpdate(src io.ReadCloser, assetURL, cmdPath string) error {
-	defer src.Close()
+func uncompressAndUpdate(src io.Reader, assetURL, cmdPath string) error {
+	if rc, ok := src.(io.ReadCloser); ok {
+		defer rc.Close()
+	}
```
[the remark this answers is not in the document]
Comment (tobiaskohlbau) [reply selfupdate/update.go]: I thought about that. Then I've to prefetch the content in `UpdateTo` as by now the http response body is passed in directly. If you're fine with moving this around I'k fine as well.

**Comment by tobiaskohlbau** [comment]
@rhysd I've addressed your comment and added samples within the README.md. Please let me know if there are more things to improve.

**Comment by rhysd** [comment]
@tobiaskohlbau 

https://github.com/rhysd/go-github-selfupdate/pull/16/files#diff-d52fa5feca807ec73fd240d0ee9e75a8R165

Could you also add `defer src.Close()` in `UpdateTo` function since closing a `ReadCloser` is now a responsibility of caller of `uncompressAndUpdate()`.

**Comment by rhysd** [comment]
> Why both functions have different signatures?

This is historical reason. At first there was no `Updater` struct. But later, it was required. To maintain API compatibility, I left `func UpdateTo()` function and added the same name method `func (u *Updateor) UpdateTo()`.

**Comment by rhysd** [comment]
I confirmed all tests are passing on my local. Thank you for contributing this! Merging.

**Pull request merged**